### PR TITLE
hints: error injection for pausing hint replay

### PR DESCRIPTION
**Pull request by piodul**
Adds a `hinted_handoff_pause_hint_replay` error injection point. When
enabled, hint replay logic behaves as if it is run, but it gets stuck in
a loop and no hints are actually sent until the point is disabled again.

This injection point will be useful in dtests - it will simulate
infinitely slow hint replay and will make it possible to test how some
operations behave while hint replay logic is running.

The first intended use case of this injection point is testing the HTTP
API for waiting for hints (#8728).

Refs: #6649

### CODE REVIEW

**Comment by piodul** [comment]
This is an alternative to https://github.com/scylladb/scylla/pull/8735

File: db/hints/manager.cc
Comment (nyh) [inline db/hints/manager.cc]: A thought for the future: would have been nice to have a way to wait for injection disable event, instead of arbitrarily polling after 100ms.

**Review by nyh — APPROVED** [review]

File: db/hints/manager.cc
Comment (kbr-) [reply db/hints/manager.cc:873]: It's not necessary if `read_log_file` keeps the lambda object alive for the whole duration of the lambda's execution.

Comment (piodul) [reply db/hints/manager.cc:873]: Yes - actually, it's not necessary for `secs_since_file_mod`, `fname` and `ctx_ptr`, either.

I had some problems with ASAN complaining when running dtests in debug mode and I mechanically applied the rename trick to all of the variables captured there, which accidentally fixed my problem. Now I see that the problem was actually a use-after-move caused by `send_one_hint(std::move(ctx_ptr), ...)` - the `ctx_ptr` captured variable was moved out and invalid in the next invocation of the lambda. I accidentally fixed it by adding the `auto ctx_ptr = ctx_ptr_;` line - the captured pointer is now copied and stays valid in the next invocation.

I'll send a v2 soon which will remove those renames.

**Comment by piodul** [comment]
v2: Rebased and removed unnecessary renames of fields from the lambda capture list.

@nyh Please do not merge until @vladzcloudius approves, too.

**Comment by piodul** [comment]
By the way, I updated the test for the HTTP API feature and pushed it [here](https://github.com/piodul/scylla-dtest/blob/61dac241a30722ea8cd2630f1e3ea6521ddbeeb0/hintedhandoff_additional_test.py#L614). I will update the PR in dtests if this PR is merged instead of #8735.

I checked that the PR #8728 passes the updated dtest after applying changes from this PR.

**Comment by piodul** [comment]
v3: Rebased - there was a tiny conflict in imports.

**Review by vladzcloudius — APPROVED** [review]
Looks good. Thanks.

**Comment by piodul** [comment]
v4: Rebased onto current master (again, there was a conflict with imports).

@nyh The HTTP API PR, which this PR was supposed to test, was merged + this PR got an approval from Vlad. May I ask you for a merge?

**Comment by piodul** [comment]
@nyh ping for merge

**Pull request closed without merging**
